src/clients/java/src/client.zig

```diff
@@ -93,415 +378,305 @@ const ReflectionHelper = struct {
         request_end_request_method_id = null;
     }
 
-    pub fn throwInitializationException(env: *jui.JNIEnv, status: tb.tb_status_t) void {
+    pub fn initialization_exception_throw(env: *jni.JNIEnv, status: tb.tb_status_t) void {
         assert(initialization_exception_class != null);
         assert(initialization_exception_ctor_id != null);
 
-        var exception = env.newObject(
+        var exception = env.new_object(
             initialization_exception_class,
             initialization_exception_ctor_id,
-            &[_]jui.jvalue{jui.jvalue.toJValue(@bitCast(jui.jint, @enumToInt(status)))},
-        ) catch {
+            &[_]jni.JValue{jni.JValue.to_jvalue(@bitCast(jni.JInt, @enumToInt(status)))},
+        ) orelse {
             // It's unexpected here: we did not initialize correctly or the JVM is out of memory.
-            @panic("JNI: Unexpected error creating a new exception.");
+            JNIHelper.vm_panic(env, "Unexpected error creating a new InitializationException.", .{});
         };
+        defer env.delete_local_ref(exception);
 
-        assert(exception != null);
-        defer env.deleteReference(.local, exception);
+        const jni_result = env.throw(exception);
+        JNIHelper.check_jni_result(
+            env,
+            jni_result,
+            "Unexpected error throwing InitializationException.",
+            .{},
+        );
 
-        env.throw(exception) catch {
-            @panic("JNI: Unexpected error throwing an exception.");
-        };
-        assert(env.hasPendingException());
+        assert(env.exception_check() == .jni_true);
+    }
+
+    pub fn assertion_error_throw(env: *jni.JNIEnv, message: [:0]const u8) void {
+        assert(assertion_error_class != null);
+
+        const jni_result = env.throw_new(assertion_error_class, message.ptr);
+        JNIHelper.check_jni_result(
+            env,
+            jni_result,
+            "Unexpected error throwing AssertionError.",
+            .{},
+        );
+        assert(env.exception_check() == .jni_true);
     }
 
-    pub fn get_send_buffer(env: *jui.JNIEnv, this_obj: jui.jobject) ?[]u8 {
+    pub fn get_send_buffer_slice(env: *jni.JNIEnv, this_obj: jni.JObject) ?[]u8 {
         assert(this_obj != null);
         assert(request_send_buffer_field_id != null);
         assert(request_send_buffer_len_field_id != null);
 
-        var buffer_obj = env.getField(.object, this_obj, request_send_buffer_field_id) orelse
+        var buffer_obj = env.get_object_field(this_obj, request_send_buffer_field_id) orelse
+            return null;
+        defer env.delete_local_ref(buffer_obj);
```

Comment:
Hm, this smells suspiciously to me, although I don't know exactly how this works, so might be wrong. 

It seems that `get_object_field` registers a refenrence with JVM, which pins the object in place and prevents GC from freeing it. That's why we need `delete_local_ref` to signal that we are done.

But we are returning a slice pointing inside the buffer! Sholdn't we delete the local ref after we are done with it? Or would it be corret to delete_local_ref here immediately, without defer, because we are holding onto the request object anyway?

Reply:
> That's why we need delete_local_ref to signal that we are done

Actually, no. Local references live on the JVM stack and are freed (decrease the Rc) automatically when a function ends. Normally only global references need to be explicitly deleted. In our case, since we run code from an unmanaged thread, we need to free both global and local references, otherwise, the JVM would not know when to pop the reference from the call stack.

> But we are returning a slice pointing inside the buffer! 

It's correct to free the buffer object reference here. The buffer's content is a stable address from unmanaged memory allocated by [UNSAFE.alloc(...)](https://github.com/openjdk/jdk21/blob/830279e0540c01e2012c60b724857a7fe4a450b1/src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template#L131), and the buffer object is only a Rc that is still being held by the Request object.

Note this is valid only for DirectBuffers, for primitive arrays and strings we need to keep the reference and pin/unpin the content.